common/src/main/java/com/microsoft/identity/common/internal/ui/webview/certbasedauth/NfcSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -51,14 +52,39 @@ public NfcSmartcardCertBasedAuthChallengeHandler(@NonNull final Activity activit
     }
 
     /**
-     * Pauses smartcard discovery, if the particular authentication method isn't meant to have
-     *  discovery active throughout the entire flow.
+     * When a connection is no longer actively being used, the dialog flow should pause
+     * so the user can remove their smartcard before flow can continue.
+     * @param callback {@link IDisconnectionCallback}
      */
     @Override
-    protected void onPausedSmartcardDiscovery() {
-        //Helps prevent unnecessary callback trigger. Nfc discovery should only be active when
-        // the user is prompted to tap.
-        mCbaManager.stopDiscovery(mActivity);
+    protected void pauseToCloseConnection(@NonNull final IDisconnectionCallback callback) {
+        if (!mCbaManager.isDeviceConnected()) {
+            callback.onClosedConnection();
+            return;
+        }
+        mDialogHolder.showSmartcardRemovalPromptDialog();
+        mCbaManager.setDisconnectionCallback(new IDisconnectionCallback() {
+            @Override
+            public void onClosedConnection() {
+                mDialogHolder.dismissDialog();
+                mCbaManager.clearDisconnectionCallback();
+                //Helps prevent unnecessary callback trigger. Nfc discovery should only be active when
+                // the user is prompted to tap.
+                mCbaManager.stopDiscovery(mActivity);
+                callback.onClosedConnection();
+            }
+        });
+        mCbaManager.disconnect();
```

Comment:
I see that we are setting the DisconnectCallback and then calling disconnect right after. Are there situations where CbaManager would call disconnect again even after the connection has been closed? If not, I think maybe we would not need to set the callback here and can just call the cancelation code:

                mDialogHolder.dismissDialog();
                mCbaManager.clearDisconnectionCallback();
                //Helps prevent unnecessary callback trigger. Nfc discovery should only be active when
                // the user is prompted to tap.
                mCbaManager.stopDiscovery(mActivity);
                callback.onClosedConnection();

Reply:
The callback is set right before disconnect() because disconnect() contains smartcard api specific logic that waits for the user to remove their smartcard, and then calls the disconnectCallback. So calling the cancellation code directly after won't really work here, but I think the equivalent of your point would be to pass the code into disconnect() as a callback parameter instead. I did that initially, but opted for setting the disconnectCallback instead, in order to keep the disconnect callback logic consistent. I could go back to passing as a callback parameter though if that would make more sense. That would essentially leave the disconnectcallback variable within the NFC manager useless, so I would have to rethink some of the shared logicof the abstract classes, and maybe add some generics back in.
I'll take some time to move stuff around again and get back to this comment.

Reply:
Coming back to this comment... I did end up adding a disconnectCallback object as a parameter within disconnect(), and the class variable mDisconnectCallback no longer exists for the NFC manager.